Subscription rework and Pedestrians

truenicfel opened the pull request:
This contains a rework of how subscriptions are handled in veins. Not yet integrated into the old code. See the commit messages for details.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+     * response for a person value.
+     *
+     * This will create new subscriptions for specific persons if the buffer
+     * contained an ID_LIST subscription.
```
Better move this into its own named method. Also: I this common behavior of all `TraciSubscriptionManager` specializations? Then it should be part of the common interface!

truenicfel replied:
This is in a different method. The comment got removed in my latest changes.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+     *
+     * @return a list of TraCIPerson.
+     */
+    std::list<TraCIPerson> getUpdated();
```
Any particular (and good) reason to use a `std::list` here? If not,just us a `std::vector`.

truenicfel replied:
modified.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+     *
+     * @return std::set<std::string> a set of person ids.
+     */
+    std::set<std::string> getDisappeared();
```
Maybe just return a `std::vector`. Using a list internally is fine, but I guess most users just need something to iterate over.

truenicfel replied:
modified.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+     * This contains all updated persons. Will be cleared after
+     * getUpdated() was called.
+     */
+    std::list<TraCIPerson> mUpdatedPersons;
```
Any particular (and good) reason to use a std::list here? If not,just us a std::vector. (As above)

truenicfel replied:
modified.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+    std::set<std::string> mDisappearedPersons;
+
+    /**
+     * Get the active person id set.
```
What exactly does *active mean*? Why is this private?

truenicfel replied:
This is a relic from the past. Forgot to remove it. Removed now.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+    void processPersonIDList(std::list<std::string>& idList);
+
+    /**
+     * Subscribe to specific TraCI person variables.
```
What are those specific variables? Can they be changed or read somewhere?
E.g., the variables could be put into a `public const static std::vector` of TraCI Variables.

truenicfel replied:
variable names added. They are all defined in TraCIConstants.h header file.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:
```diff
@@ -0,0 +1,267 @@
+            // now check which variable id we got
+            // it is either an id_list or a subscription for a specific vehicle
+            // never both
+            if (responseVariableID == TraCIConstants::ID_LIST) {
```
This can be simplified a lot if you move handling `TraCIConstants::ID_LIST`s to it's own named method.

truenicfel replied:
That is done a few lines below.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:
```diff
@@ -0,0 +1,267 @@
+
+std::list<TraCIPerson> PersonSubscriptionManager::getUpdated() {
+    std::list<TraCIPerson> temp = mUpdatedPersons;
+    mUpdatedPersons.clear();
```
Why does this getter have to modify the state of the Manager? Why not just return a (copied) vector of updated persons? The clearing could be done in the update or explicitly.

truenicfel replied:
i changed it so that the clearing has to be done explicitly from outside.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:
```diff
@@ -0,0 +1,267 @@
+
+std::set<std::string> PersonSubscriptionManager::getDisappeared() {
+    std::set<std::string> temp = mDisappearedPersons;
+    mDisappearedPersons.clear();
```
Why does this getter have to modify the state of the Manager? Why not just return a (copied) vector of disappeared persons? The clearing could be done in the update or explicitly.

truenicfel replied:
see above.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+     * Process a subscription result that contains an id list. This is
+     * simply a helper method to improve clarity of update().
+     */
+    void processPersonIDList(std::list<std::string>& idList);
```
This method seems to make sure that the active subscriptions match `idList`, so why not give it a more explanatory name?

truenicfel replied:
It compares active subscriptions with the given id_list and adds disappeared persons to disappeared vector and new person to updated vector.

I can't think of a more explanatory name but if you suggest something I'm happy to edit.

dbuse replied:
Can't really think of a better name right now. Maybe @mightyCelu has an idea :D

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCISubscriptionManager.h:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * Default constructor.
+     *
+     * @param explicitUpdateIfSubscriptionUnavailable In case there is no subscription response
```
I don't understand what this means...

truenicfel replied:
This controls if an explicit update for person and vehicle manager should be performed if they did not receive an id_list subscription. See the comment above where I explained why I need the `updateWithList` method in person subscription manager.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCISubscriptionManager.h:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * Initialize this manager with the given parameters to access the traci server.
+     */
+    void initialize(std::shared_ptr<TraCIConnection> connection, std::shared_ptr<TraCICommandInterface> commandInterface);
```
As with the individual Managers, move this into the constructor an you never have to worry about uninirialized instances.

truenicfel replied:
see discussion above.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCISubscriptionManager.h:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * Default destructor.
+     */
+    virtual ~TraCISubscriptionManager() = default;
```
This class seems like it is not intended to be used polymorphically, so you can just omit this.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCISubscriptionManager.h:
```diff
@@ -0,0 +1,218 @@
+     *
+     * @return list<TraCIPerson> a list of persons.
+     */
+    std::list<TraCIPerson> getUpdatedPersons();
```
This and all other getters should just return a copy (idally a vector) and be const.

mightyCelu replied:
Also you might want to consider using `std::vector`, there are few use-cases where a `std::list` should be preferred.

truenicfel replied:
done.

dbuse replied:
Alright.

dbuse commented on src/veins/modules/mobility/traci/subscriptionManagement/TrafficLightSubscriptionManager.h:
```diff
@@ -0,0 +1,91 @@
+     *
+     * @param id of the traffic light to subscribe to.
+     */
+    void subscribeToTrafficLight(std::string id);
```
Maybe rename to `subscribe`, this Manager only considers traffic lights.

truenicfel replied:
done.

dbuse replied:
Alright.

truenicfel commented:
Thanks for the review! How would you like the changes implemented? New commits on top of the old ones or modifications to existing commits?

dbuse commented:
I think its easier if you modify the commits. That way you could also integrate the code formatting back into them. Also, there is only one person working on the code, so rebasing/force-pushing should not be too much of  a problem.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/TrafficParticipant.h:
```diff
@@ -0,0 +1,145 @@
+ *
+ * @author Nico Dassler <dassler@hm.edu>
+ */
+class TrafficParticipant {
```
changed to struct. getters also removed.

dbuse replied:
Great!
Personally, I'm not a big fan of prefixing all field names with an `m` and can't think of any other place in Veins where that is done, but that should not be a blocker.

truenicfel replied:
I removed all the `m` prefixes. They are a habit of mine from the time when I first started programming c++. Should have looked for conventions in the project first.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCIPerson.h:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Nico Dassler <dassler@hm.edu>
+ */
+class TraCIPerson: public TrafficParticipant {
```
changed to struct. getters also removed.

dbuse replied:
I would keep this as a specialized class, so we can distinguish the type.
```suggestion
struct Person: public TrafficParticipant {};
```

truenicfel replied:
I'm sorry I don't understand :c

The state right now is that there are 3 structs. TrafficParticipant being the parent with two children (Person and Vehicle). Do you suggest all of them being classes or all of them being structs (or some kind of mixture)?

dbuse replied:
First, all of them should be structs.
Second, even `Person` (which does not add any fields by itself) should be a struct inheriting from `TrafficParticipant`, not just a typedef.

truenicfel replied:
ok got it. thanks. changed.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCIVehicle.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * @author Nico Dassler <dassler@hm.edu>
+ */
+class TraCIVehicle: public TrafficParticipant {
```
changed to struct. getters also removed.

dbuse replied:
Great (same as with TrafficParticipant).

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/TraCITrafficLight.h:
```diff
@@ -0,0 +1,81 @@
+    std::string currentProgram;
+
+    /**
+     * The assumed time (in seconds) at which the tls changes
```
i can remove that but as stated in the comment this is just copied from the official documentation :P
--> https://sumo.dlr.de/wiki/TraCI/Traffic_Lights_Value_Retrieval

dbuse replied:
True, but only since SUMO 1.0. We actually jumped through some loops to keep Veins compatible with older SUMO versions, where times in TraCI were still in milliseconds.
Also, time unit is abstracted away by `simtime_t`

truenicfel replied:
Removed.

dbuse replied:
Great.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     * @param connection to access traci server (for example: to perform subscriptions).
+     * @param commandInterface to access traci (for example: direct request to get vehicle id_list)
+     */
+    virtual void initialize(std::shared_ptr<TraCIConnection> connection, std::shared_ptr<TraCICommandInterface> commandInterface);
```
When I change the value of a shared_ptr does that change become visible in all classes holding that pointer?

mightyCelu replied:
Yes, the object behind the `shared_ptr` is the same for all if its instances. In existing code however, we used regular pointers to indicate that they are non-owning. The change to `shared_ptr` seems like it does not reflect the semantics properly - the connection's owner is the `TraCIScenarioManager`, and also it is incompatible with the current code base where, for that reason, we use a `unique_ptr`. Is there a particular reason you opted to change this to `shared_ptr`s?

A less invasive integration with the `TraCIScenarioManager` would be to use its `traci_initialized` signal, this way you can avoid integrating it deeply within it.
However, since the subscription managers it might be worth putting some thought into what constitutes TraCI being initialized: Does this require that the subscriptions are being set up? If so, we might want to add TraCI pre-, post-, and actual-initialization signals.

truenicfel replied:
Hm. I used a `shared_ptr` because I thought (might be wrong) that it is not possible to simply pass around a `unique_ptr`. (I'm not sure but I think my expectation got proven correct when I got a compilation error. Mabe used it wrong...).

The reason why I need to pass these pointers is because the different managers need access to them to perform subscriptions.

In my opinion the best solution would be a "global" object that is instantiated in the manager and then automatically shared with all managers. Then there will be a call to a method (something like `setUpSubscriptions`) that all managers implement which set ups the basic subscriptions.

I can't really comment on the second paragraph since i do not understand what it means "to use its `traci_initialized` signal". (I'm sorry I'm still a beginner with omnet++).

dbuse replied:
I think the easiest and cleanest way to solve (or get around) this problem is to only construct all subscription managers once the `TraCIConnection` is established (and subsequently, `TraCICommandInterface` is available, too). Then, we can get rid of this whole `initialize` method for the subscription managers and they cannot get into an uninitialized state.

Whether to use a raw or shared pointer (or even a reference, which is possible if we pass it in at construction, and thus would be my suggestion) is still open, but only a minor issue.

So I would do the following:
- Once TraCI is initialized (which can be detected by the `TraCIScenarioManager::traci_initialized` simulation signal (see [the omnet++ manual about signals](https://doc.omnetpp.org/omnetpp/manual/#sec:simple-modules:signals) and [some info about using those signals in veins](https://github.com/sommer/veins/blob/master/doc/manual/simsignal_management.md)):
- The module owning the `SubscriptionManager` (probably the `TraCIScenarioManager`) constructs the `SubscriptionManager`, passing in the `TraCIConnection` and `TraCICommandInterface` (by reference or pointer). As this happens after the construction of the owning module, it should be constructed into a `std::unique_ptr`.
- The `SubscriptionManager` in turn constructs the individual subscription managers, handing on `TraCIConnection` and `TraCICommandInterface` references/pointers. The individual subscription managers are just constructed as data members (no pointers) and valid/initialized immediately after construction.
- The whole system is ready and can start processing subscriptions.

mightyCelu replied:
@dbuse I like the design you describe.
@truenicfel You're right, you cannot create copies of a `unique_ptr` around as it has ownership semantics. However, you can access the underlying pointer and treat it as a 'weak pointer', i.e., a reference that will not keep the object alive.

truenicfel replied:
So here is what I did now:

- `unique_ptr<ExecutiveSubscriptionManager> subscriptionManager;` as class variable in `TraCIScenarioManager` which is initialized in the `init_traci()` method of the scenario manager
- two shared ptrs for command interface and connection in `TraCIScenarioManager` 
- the two shared ptrs will be passed (by reference) into the constructor of everything related to subscriptions

Please note that I will push the changes to `TraCIScenarioManager` when we are finished with every file in directory `.../subscriptionManagement` (see comment below).

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @return true if adding was successful.
+     */
+    bool addToSubscribed(std::string id);
```
Name:
I did not call it subscribe because there is no subscription performed here. The id is only added to a set of subscribed ids. (The comment above the method is misleading though. I will change that.)

Exception + return value:
I will throw an exception when a duplicate got added.

dbuse replied:
All clear about the error handling.

For the name and semantics: Is there a way to register some object for subscription then? Or can that only be done via `ID_LIST` results?

truenicfel replied:
That depends on the manager. The methods defined in the base class of each manager are just for tracking which ids have been subscribed and are there to reduce the amount of duplicate code I wrote.

The different managers implement their own methods to subscribe to a specific id (for example persons or vehicles) at the traci server. For example: https://github.com/truenicfel/veins/blob/subsRework/src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc#L231
*(This is from my fork of veins)*

The `ID_LIST` subscription tells me which ids of cars are currently active. The container in the manager base class will get updated with these infos. The newly appeared ids will then be fed
into a method similar to the one linked above.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @return true if removing successful.
+     */
+    bool removeFromSubscribed(std::string id);
```
See above.

dbuse replied:
Alright.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @return set of subscribed ids.
+     */
+    std::set<std::string> getSubscribed();
```
I updated it to use a vector.

dbuse replied:
Great.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @return shared pointer to TraCIConnection.
+     */
+    std::shared_ptr<TraCIConnection> getConnection();
```
I changed some of these methods to protected. So they can be accessed from child class but not from outside.

dbuse replied:
Alright.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *
+     * @return shared pointer to TraCICommandInterface.
+     */
+    std::shared_ptr<TraCICommandInterface> getCommandInterface();
```
see above.

dbuse replied:
Alright.

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+namespace TraCISubscriptionManagement {
+
+/**
+ * @class SubscriptionManagerBase
```
done.

dbuse replied:
Thanks!

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     *      - the response command length extended
+     *      - the id of the response commmand
+     *
+     * This method is expected to change the results to api methods of this manager.
```
modified.

dbuse replied:
thanks

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+     * is assumed to have the first few bytes removed. These bytes represent
+     * the first few fields of an answer to a simulation step request or the
+     * request to subscribe variable request.
+     * The fields that should be removed are:
```
modified.

dbuse replied:
thanks

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/SubscriptionManagerBase.h:
```diff
@@ -0,0 +1,162 @@
+    virtual ~SubscriptionManagerBase() = default;
+
+    /**
+     * Each subscription manager has to implement this method. The given buffer
```
done

dbuse replied:
thanks

truenicfel commented on src/veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.h:
```diff
@@ -0,0 +1,153 @@
+    virtual ~PersonSubscriptionManager() = default;
+
+    /**
+     * Update this PersonSubscriptionManager with the given
```
I improved the documentation this does the same as the update method when it receives an ID_LIST subscription result. 

A little bit of history: When I first started working on this I could not get the person subscription working. The initial subuscription acknowledged by traci but i never received any subscription results. Therefore I had to do an explicit update. This explicit upate found its way into this method.

truenicfel commented:
Phew. That was a lot! I updated everything now.

The main changes that I made:

- clearAPI(): this method will allow to clear the api of managers from outside. so that multiple calls to getters in managers will not return different results.
- removed traCI from names
- modified traffic participants to be structs
- const to getters added
- copyright adapted (hopefully all)
- #include statements added and sorted (hopefully all)

Thats about it. The main issue that is not resolved yet are the `initialize` methods where I could not think of a proper solution (yet). See discussion above.

I tested the whole thing again with my example and the testsimulation in the subprojects folder and it still works!

dbuse commented:
When I try to compile this, I am getting some errors:
```
veins/modules/mobility/traci/subscriptionManagement/ExecutiveSubscriptionManager.cc:66:55: error: no member named 'RESPONSE_SUBSCRIBE_PERSON_VARIABLE' in namespace 'veins::TraCIConstants'
        else if (responseCommandID == TraCIConstants::RESPONSE_SUBSCRIBE_PERSON_VARIABLE) {
                                      ~~~~~~~~~~~~~~~~^
veins/modules/mobility/traci/subscriptionManagement/ExecutiveSubscriptionManager.cc:83:9: error: no type named 'Person' in 'veins::TraCICommandInterface'; did you mean simply 'Person'?
        veins::TraCICommandInterface::Person defaultPerson = mCommandInterface->person("");
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        Person
./veins/modules/mobility/traci/subscriptionManagement/Person.h:38:28: note: 'Person' declared here
typedef TrafficParticipant Person;
                           ^
veins/modules/mobility/traci/subscriptionManagement/ExecutiveSubscriptionManager.cc:83:81: error: no member named 'person' in 'veins::TraCICommandInterface'
        veins::TraCICommandInterface::Person defaultPerson = mCommandInterface->person("");
                                                             ~~~~~~~~~~~~~~~~~  ^
veins/modules/mobility/traci/subscriptionManagement/ExecutiveSubscriptionManager.cc:90:56: error: no member named 'getIdList' in 'veins::TraCICommandInterface::Vehicle'
        std::list<std::string> idList = defaultVehicle.getIdList();
                                        ~~~~~~~~~~~~~~ ^
3 warnings and 4 errors generated.
make[1]: *** [Makefile:212: ../out/clang-debug/src/veins/modules/mobility/traci/subscriptionManagement/ExecutiveSubscriptionManager.o] Error 1
make[1]: *** Waiting for unfinished jobs....
veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:190:62: error: no member named 'CMD_SUBSCRIBE_PERSON_VARIABLE' in namespace 'veins::TraCIConstants'
    TraCIBuffer buf = getConnection()->query(TraCIConstants::CMD_SUBSCRIBE_PERSON_VARIABLE, TraCIBuffer() << beginTime << endTime << objectId << variableNumber << variable1);
                                             ~~~~~~~~~~~~~~~~^
veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:201:49: error: no member named 'RESPONSE_SUBSCRIBE_PERSON_VARIABLE' in namespace 'veins::TraCIConstants'
    ASSERT(responseCommandID == TraCIConstants::RESPONSE_SUBSCRIBE_PERSON_VARIABLE);
                                ~~~~~~~~~~~~~~~~^
/home/dbuse/.checkouts/manual/omnetpp/current/include/omnetpp/simkerneldefs.h:56:13: note: expanded from macro 'ASSERT'
  ((void) ((expr) ? 0 : (throw omnetpp::cRuntimeError("ASSERT: Condition '%s' does not hold in function '%s' at %s:%d", \
            ^~~~
veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:250:65: error: no member named 'CMD_SUBSCRIBE_PERSON_VARIABLE' in namespace 'veins::TraCIConstants'
    TraCIBuffer buffer = getConnection()->query(TraCIConstants::CMD_SUBSCRIBE_PERSON_VARIABLE, TraCIBuffer() << beginTime << endTime << objectId << variableNumber << variable1 << variable2 << variable3 << variable4 << variable5);
                                                ~~~~~~~~~~~~~~~~^
veins/modules/mobility/traci/subscriptionManagement/PersonSubscriptionManager.cc:261:49: error: no member named 'RESPONSE_SUBSCRIBE_PERSON_VARIABLE' in namespace 'veins::TraCIConstants'
    ASSERT(responseCommandID == TraCIConstants::RESPONSE_SUBSCRIBE_PERSON_VARIABLE);

```

dbuse commented:
Went through the updated code again. Thanks for the fast response, we are really moving forward with this.
I've added comments to all the discussions above. See especially the comment to the construct/initialize discussion.

truenicfel commented:
> When I try to compile this, I am getting some errors:
`...`

I'm sorry for that. I made a mistake there. I opened a pull request a few days ago (https://github.com/sommer/veins/pull/147). I got feedback to split it up because it was quite a lot of changes. I split it up into two big parts. Changes that do not modify existing code and changes that modify existing code. 

I forgot that there are dependencies from the changes in this pull request to the other one. These are the changes to the TraCIConstants header file. The used constants are undefined. That is why it is not possible to compile. I will add a commit later to include the TraCIConstants file with this pull request. After all we don't want a master branch that is not possible to compile.

truenicfel commented:
I have updated the commits again.

Main changes this time:
- `initialize()` removed (see comments above)
- `m` prefixes removed

truenicfel commented:
So I updated the commits once more. I changed the person typedef to a struct. It seems to me that there are no more open discussions. As soon as you guys are ready I will push the remaining changes to my branch to integrate the changes into the existing code.

truenicfel commented:
Since nothing happened here for a couple of days I decided to add the last commits which make modifications to `TraCIScenarioManager`, `TraCICommandInterface` and `TraCIConstants`. I also added a small example which does nothing except showing the visuals of pedestrians displayed in sumo. It is in the `subprojects/veins_inet` subproject. 

Everything should build now and I rebased everything with recent changes to master branch. I ran the `veins_testsims` example as well as the Erlangen example to ensure that everything is still working as intended.

Please feel free to contact me via mail in case I forget to check back for new comments here.

dbuse commented:
Thanks again. I'll try to look into the latest additions in the next couple of days.

sommer commented:
I truly appreciate all the hard work that has gone into this and I wish I could merge the pull-request as-is, but I just cannot. Since last commenting here (<https://github.com/sommer/veins/pull/147#issuecomment-472450820>), I have repeatedly tried to understand what changes this pull request introduces, but failed each time. A large part of this is due to many commits seemingly contributing new code by you (e.g., <https://github.com/sommer/veins/pull/148/commits/a78676dde1f5f9eb2f71acbcac778a90385b137d#diff-d4c2c2bdd18de556d5d43691bb2ae271R257>) when in reality this code is just copy&pasted from old code (in this case, from <https://github.com/sommer/veins/blob/veins-5a1/src/veins/modules/mobility/traci/TraCIScenarioManager.cc#L588>), which a later commit in the pull request then removes. In this instance, it is easy enough for a determined reader to figure out that a vehicle now subscribes to one more variable (`VAR_TYPE`), but this should really be obvious from a commit (aside from being a documented change, not a side effect of a new architecture). Other changes I found include the removal of the `isUsable()` method that Plexe critically depends on (<https://github.com/michele-segata/plexe-veins/blob/d44f0518143cb23362044e2bc90ca90b8dfe0736/src/plexe/PlexeManager.cc#L16>) as well a much more fundamental change from tracking vehicle departures/arrivals/teleporting and acting accordingly, now instead assuming that every vehicle that SUMO does not report in its subscription list should be destroyed (and, vice versa, if a vehicle remains in a subscription list it should not be destroyed), both of which might be flawed. All of this straight-out bars me from merging. Still, the Person code itself is done very well, so I hope this pull request can serve as a valuable resource for how to integrate Persons (or similar SUMO objects) into Veins to anyone who might be interested.
